### PR TITLE
[DataPipe] Control shuffle settings from DataLoader2

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Makes `shuffle` DataPipe sensitive to DataLoader(2) `shuffle` kwarg.

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -224,11 +226,14 @@ def __init__(self, dataset: Dataset[T_co], batch_size: Optional[int] = 1,
             # option. If this turns out to be useful in future, we can re-enable
             # this, and support custom samplers that specify the assignments to
             # specific workers.
-            if shuffle is not False:
+            if isinstance(dataset, IterDataPipe):
+                torch.utils.data.graph_settings.apply_shuffle_settings(dataset, shuffle=shuffle)
+            elif shuffle is not False:
```
Comment (ejguan) [inline torch/utils/data/dataloader.py:231]: The branch that `shuffle` is not False will never be reached.
Could we add a return boolean value for `apply_shuffle_settings` function to check if `dataset` / data pipeline has at least one `Shuffler`?
If false, raise the following ValueError.

File: test/test_dataloader.py [hunk 2]
```diff
@@ -1132,10 +1132,6 @@ def test_invalid_ctor_args_combinations(self):
 
         # iterable-style
         dataset = CountingIterableDataset(20)
-        with self.assertRaisesRegex(ValueError, "DataLoader with IterableDataset: expected unspecified shuffle"):
-            self._get_data_loader(dataset, shuffle=True)
-        with self.assertRaisesRegex(ValueError, "DataLoader with IterableDataset: expected unspecified shuffle"):
-            self._get_data_loader(dataset, shuffle=3)
```
Comment (ejguan) [inline test/test_dataloader.py:1138]: Then, this part can be preserved.

**Review by ejguan — APPROVED** [review]
LGTM with one comment below.

File: torch/utils/data/dataloader.py (hunk 1 above)
Comment (VitalyFedyunin) [reply torch/utils/data/dataloader.py:231]: Addressing both comments. It is not the first time when `isinstance(dp, IterDataPipe)` not doing the proper job. I think it is better if we untie IterDataPipe and IterDataset with something simple like:

```
class IterDataPipe(IterDataset):
  pass
```

Comment (ejguan) [reply torch/utils/data/dataloader.py:231]: SGTM!!

File: torch/utils/data/datapipes/iter/combinatorics.py [hunk 3]
```diff
@@ -86,6 +87,7 @@ def __init__(self,
         else:
             self.datapipe = datapipe.unbatch(unbatch_level=unbatch_level)
         self.buffer_size = buffer_size
+        self._shuffle_enabled = True
```
Comment (pmeier) [inline torch/utils/data/datapipes/iter/combinatorics.py:90]: As discussed in our latest sync, please add an option to set this when creating the `Shuffler`. That way we can include the `Shuffler` in "the right spot" in the pipeline without actually shuffling until explicitly requested.

File: torch/utils/data/dataset.py [hunk 4]
```diff
@@ -219,9 +215,15 @@ def __add__(self, other: Dataset[T_co]):
     # No `def __len__(self)` default? Subclasses raise `TypeError` when needed.
     # See NOTE [ Lack of Default `__len__` in Python Abstract Base Classes ]
 
+
+class IterDataPipe(IterableDataset[T_co], metaclass=_DataPipeMeta):
+    functions: Dict[str, Callable] = {}
+    reduce_ex_hook : Optional[Callable] = None
+    getstate_hook: Optional[Callable] = None
```
Comment (ejguan) [inline torch/utils/data/dataset.py:222]: Yeah, this is awesome! Just a note that we may need to change here https://github.com/pytorch/pytorch/blob/29d759948ee7cd2275ea60ebd62c311408e3798b/torch/utils/data/__init__.py#L19

Comment (ejguan) [reply torch/utils/data/dataset.py:222]: NVM, I didn't notice you have already changed it.

File: torch/utils/data/graph.py [hunk 5]
```diff
@@ -41,18 +41,21 @@ def reduce_hook(obj):
             return stub_unpickler, ()
 
     # TODO(VitalyFedyunin):  Better do it as `with` context for safety
-    IterableDataset.set_reduce_ex_hook(reduce_hook)
+    IterDataPipe.set_reduce_ex_hook(reduce_hook)
```
Comment (NivekT) [inline torch/utils/data/graph.py:44]: Would using the `try... finally` pattern help with safety? Or do we have to use `with`?

```python
try:
    IterDataPipe.set_reduce_ex_hook(reduce_hook)
    if exclude_primitive:
        IterDataPipe.set_getstate_hook(getstate_hook)
    p.dump(scan_obj)
except AttributeError:  # unpickable DataPipesGraph
    pass  # TODO(VitalyFedyunin): We need to tight this requirement after migrating from old DataLoader
finally:
    IterDataPipe.set_reduce_ex_hook(None)
    if exclude_primitive:
        IterDataPipe.set_getstate_hook(None)
```

**Pull request closed without merging**